### PR TITLE
Fix floating point arithmetic

### PR DESCRIPTION
**Pull request by AnanyaSshri**
This PR fixes the usage of floating point arithmetic in LIPs by replacing `floor` or `ceil` with the following:

- `floor(x / y )` is replaced with `x // y`
- `ceil(x / y)` is replaced with `(x + y -1) // y`

where // is the integer division.

### CODE REVIEW

**Review by vjaiman — APPROVED** [review]
Looks good.

File: proposals/lip-0038.md [hunk 1]
```diff
@@ -84,7 +84,7 @@ Each aggregate signature needs to be accompanied by some information that specif
 
 ```python
 def createAggSig(keysList: list[bytes], pubKeySignaturePairs: list[tuple[bytes, bytes]]) -> tuple[bytes, bytes]:
-    aggregationBits = byte string of length ceil(length(keyList)/8) with all bytes set to 0
+    aggregationBits = byte string of length ((length(keysList)+8-1) //8) with all bytes set to 0
```
Comment (janhack) [inline proposals/lip-0038.md:87]:
```suggestion
    aggregationBits = byte string of length (length(keysList) + 7)  // 8 with all bytes set to 0
```

File: proposals/lip-0047.md [hunk 2]
```diff
@@ -610,7 +610,7 @@ Let `genesisBlockAssetBytes` be the bytes included in the `data` property of a d
   * Check that the `weight` property of every entry in the `assetPoA.snapshotSubstore.activeValidators` array is a positive integer.
   * Let `totalWeight` be the sum of the `weight` property of every entry in `assetPoA.snapshotSubstore.activeValidators`. Then, check that `totalWeight` is less than or equal to `MAX_UINT64`.
   * Check that the value of `assetPoA.snapshotSubstore.threshold` is within the following range:
-     * Minimum value: ⌊ ⅓ × `totalWeight`⌋ + 1, where ⌊⋅⌋ is the floor function.
+     * Minimum value: (`totalWeight` // 3) + 1, where // is the integer division.
```
Comment (janhack) [inline proposals/lip-0047.md:613]:
```suggestion
     * Minimum value: `totalWeight // 3 + 1`, where // is the integer division.
```

File: proposals/lip-0058.md [hunk 3]
```diff
@@ -168,11 +168,11 @@ validatorSchema = {
...
+* `prevoteThreshold`: This property stores the prevote threshold. For casting a precommit for a block, the sum of BFT weights of all validators casting a prevote for this block has to be at least the prevote threshold. If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the property value must be `(2*aggregateBFTWeight(h) // 3)+1`.
 * `precommitThreshold`: This property stores the precommit threshold. For considering a block final, the sum of BFT weights of all validators casting a precommit for this block has to be at least the precommit threshold.
-If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `floor(1/3*aggregateBFTWeight)+1 <= precommitThreshold <= aggregateBFTWeight`.
+If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `(aggregateBFTWeight // 3)+1 <= precommitThreshold <= aggregateBFTWeight`.
```
Comment (janhack) [inline proposals/lip-0058.md:173]:
```suggestion
If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `(aggregateBFTWeight // 3) + 1 <= precommitThreshold <= aggregateBFTWeight`.
```

File: proposals/lip-0058.md [hunk 4]
```diff
@@ -168,11 +168,11 @@ validatorSchema = {
...
 * `certificateThreshold`: This property stores the certificate threshold. For considering a certificate or aggregate commit for a valid block, the sum of BFT weights of all validators signing the certificate has to be at least the certificate threshold.
-If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `floor(1/3*aggregateBFTWeight)+1 <= certificateThreshold <= aggregateBFTWeight.`
+If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `(aggregateBFTWeight // 3)+1 <= certificateThreshold <= aggregateBFTWeight.`
```
Comment (janhack) [inline proposals/lip-0058.md:175]:
```suggestion
If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the value of this property must satisfy `(aggregateBFTWeight // 3) + 1 <= certificateThreshold <= aggregateBFTWeight.`
```

File: proposals/lip-0063.md [hunk 5]
```diff
@@ -541,7 +541,7 @@ def createStakersArray():
 def createGenesisDataObj():
     genesisDataObj = {}
     genesisDataObj.initRounds = POS_INIT_ROUNDS
-    r = ceiling((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK) / ROUND_LENGTH)
+    r = ((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK) + ROUND_LENGTH - 1) // ROUND_LENGTH
```
Comment (janhack) [inline proposals/lip-0063.md:544]: As it is not obvious that this expression computes the ceiling, I would introduce a dedicated function for it similar to LIP 57 https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#ceiling.

File: proposals/lip-0038.md (hunk 1 above)
Comment (janhack) [reply proposals/lip-0038.md:87]: Also consider adding a dedicated function for `ceiling` similar to LIP 57 https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#ceiling.

**Review by vjaiman — APPROVED** [review]
Looks good.

File: proposals/lip-0038.md [hunk 6]
```diff
@@ -125,6 +126,19 @@ def verifyWeightedAggSig(keysList: list[bytes], aggregationBits: bytes, signatur
...
+
+Returns the ceiling of the division between two positive integers.
+
+```python
+def ceiling(x: uint32,y: uint32) -> uint32:
+    if y == 0:
+        raise Exception('Can not divide by 0.')
+    return (x + y - 1) // y
+```
+
```
[the remark this answers is not in the document]
Comment (janhack) [reply proposals/lip-0038.md:141]: In general, the internal function section only is used in LIPs defining a module. In this LIP here anything could be considered an internal function as it defines rather a library. Therefore, I would maybe rename the section to "Auxiliary Function". 

Further note that the rationale can before or after the specifications. The function should be defined in the specification section though.

[the remark this answers is not in the document]
Comment (AnanyaSshri) [reply proposals/lip-0038.md:141]: Renamed it as per the suggestion.

File: proposals/lip-0063.md [hunk 7]
```diff
@@ -541,13 +541,27 @@ def createStakersArray():
...
+#### ceiling
+
+Returns the ceiling of the division between two positive integers.
+
+```python
+def ceiling(x: uint32,y: uint32) -> uint32:
+    if y == 0:
+        raise Exception('Can not divide by 0.')
+    return (x + y - 1) // y
+```
```
[the remark this answers is not in the document]
Comment (AnanyaSshri) [reply proposals/lip-0063.md:564]: Renamed it as per the suggestion.

File: proposals/lip-0063.md [hunk 8]
```diff
@@ -541,13 +541,27 @@ def createStakersArray():
 def createGenesisDataObj():
     genesisDataObj = {}
     genesisDataObj.initRounds = POS_INIT_ROUNDS
-    r = ceiling((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK) / ROUND_LENGTH)
+    r = ceiling((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK), ROUND_LENGTH)
```
Comment (janhack) [inline proposals/lip-0063.md:544]:
```suggestion
    r = ceiling(a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK, ROUND_LENGTH)
```

File: proposals/lip-0058.md [hunk 9]
```diff
@@ -168,11 +168,11 @@ validatorSchema = {
 
 All properties below are valid starting from the height given as key:
 
-* `prevoteThreshold`: This property stores the prevote threshold. For casting a precommit for a block, the sum of BFT weights of all validators casting a prevote for this block has to be at least the prevote threshold. If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the property value must be `floor(2/3*aggregateBFTWeight(h))+1`.
+* `prevoteThreshold`: This property stores the prevote threshold. For casting a precommit for a block, the sum of BFT weights of all validators casting a prevote for this block has to be at least the prevote threshold. If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the property value must be `(2 * aggregateBFTWeight(h) // 3) + 1`.
```
Comment (janhack) [inline proposals/lip-0058.md:171]:
```suggestion
* `prevoteThreshold`: This property stores the prevote threshold. For casting a precommit for a block, the sum of BFT weights of all validators casting a prevote for this block has to be at least the prevote threshold. If `aggregateBFTWeight` is the sum of BFT weights of all validators stored, then the property value must be `((2 * aggregateBFTWeight(h)) // 3) + 1`.
```

Comment (janhack) [reply proposals/lip-0058.md:171]: Here the order of operations could matter.

File: proposals/lip-0063.md (hunk 8 above)
Comment (janhack) [reply proposals/lip-0063.md:544]: Brackets around one input parameter of a function are not needed.

**Review by janhack — CHANGES_REQUESTED** [review]
I added two minor suggestions.

**Review by janhack — APPROVED** [review]

**Pull request merged**